### PR TITLE
fix: bgp.ipv*.nexthop keys treated as unknown

### PR DESCRIPTION
**Pull request by mandrav**
These keys were added to validation rules in `driver_bridge.go` instead of inside `bgpValidationRules()` in `driver_common.go`, leading the latter to declare them as "invalid network keys".

### CODE REVIEW

File: lxd/network/driver_common.go [hunk 1]
```diff
@@ -578,11 +578,23 @@ func (n *common) handleDependencyChange(netName string, netConfig map[string]str
...
 
+		switch k {
+		case "bgp.ipv4.nexthop":
+			rules[k] = validate.Optional(validate.IsNetworkAddressV4)
+		case "bgp.ipv6.nexthop":
+			rules[k] = validate.Optional(validate.IsNetworkAddressV6)
+		}
+
+		// BGP keys have the peer name in their name, extract the suffix.
+		if !strings.HasPrefix(k, "bgp.peers.") {
```
Comment (roosterfish) [inline lxd/network/driver_common.go:594]: Probably we should also return an error here if `k` doesn't have prefix `bgp.peers.`. Currently it allows setting arbitrary config keys like `bgp.abc.{peer}.holdtime`.

You can add the condition to the if statement below:

```go
if len(fields) != 4 || !strings.HasPrefix(k, "bgp.peers.") {
    return nil, fmt.Errorf("Invalid network configuration key: %q", k)
}
```

Comment (roosterfish) [reply lxd/network/driver_common.go:594]: Please quote the string with %q: `fmt.Errorf("Invalid network configuration key: %q", k)`

File: lxd/network/driver_common.go [hunk 2]
```diff
@@ -601,6 +613,9 @@ func (n *common) bgpValidationRules(config map[string]string) (map[string]func(v
 			rules[k] = validate.Optional(validate.IsAny)
 		case "holdtime":
 			rules[k] = validate.Optional(validate.IsInRange(9, 65535))
+		default:
+			// other, unknown bgp config key
+			return nil, fmt.Errorf("Invalid network configuration key: %s", k)
```
Comment (roosterfish) [inline lxd/network/driver_common.go:618]:
```suggestion
			return nil, fmt.Errorf("Invalid network configuration key: %q", k)
```

**Review by roosterfish — CHANGES_REQUESTED** [review]
Thanks for this. Looking good, just a few comments.

File: lxd/network/driver_common.go (hunk 1 above)
Comment (mandrav) [reply lxd/network/driver_common.go]: Done and done

**Comment by roosterfish** [comment]
The changes look good to me! 

Please also squash the first and last commit and prefix the messages with `lxd/network: ` so that it is easier for us to identify the changes in the commit. You can find some more info here: https://github.com/canonical/lxd/blob/main/CONTRIBUTING.md#commit-structure

**Comment by roosterfish** [comment]
There could be one commit changing the log message in `drivers_common.go` and one aligning the validation (moving over the rules and checking for bpg.peers.).

**Comment by mandrav** [comment]
Squashing and prefixing the message with `lxd/network` done.
Thanks for helping!

**Comment by roosterfish** [comment]
There are five commits now instead of two. You can keep `98599cacf92edb7f0cdb13c5295806b9949ecb89` but its message needs to be prefixed with for example `lxd/network:`. `a707e144ecb9b7221e2d008205cb047c37552ffa` looks ok. The other three commits can be removed since the changes are already contained in the other two.

**Comment by mandrav** [comment]
I think it's ok now?

File: lxd/network/driver_common.go [hunk 3]
```diff
@@ -578,15 +578,22 @@ func (n *common) handleDependencyChange(netName string, netConfig map[string]str
...
 	for k := range config {
-		// BGP keys have the peer name in their name, extract the suffix.
+		// only BGP keys are handled here.
 		if !strings.HasPrefix(k, "bgp.") {
 			continue
 		}
 
-		// Validate remote name in key.
+		switch k {
+		case "bgp.ipv4.nexthop":
```
Comment (roosterfish) [inline lxd/network/driver_common.go:587]: We also need to `continue` the loop here if either bgp.ipv4.nexthop or bgp.ipv6.nexthop.

**Review by roosterfish — CHANGES_REQUESTED** [review]

**Review by roosterfish — APPROVED** [review]
LGTM!

**Review by tomponline — CHANGES_REQUESTED** [review]
The documentation explicitly says these settings are for bridge networks only, so I dont think making them available for all bgp compatible networks is appropriate:

https://documentation.ubuntu.com/lxd/en/latest/howto/network_bgp/#configure-next-hop-bridge-only

**Comment by roosterfish** [comment]
I see. If you read through the description of the `bgp_network` API extension it gives you the impression that the `.nexthop` keys are valid for both `bridge` and `physical` networks: https://documentation.ubuntu.com/lxd/en/latest/api-extensions/#network-bgp

File: lxd/network/driver_common.go [hunk 4]
```diff
@@ -578,15 +578,24 @@ func (n *common) handleDependencyChange(netName string, netConfig map[string]str
 func (n *common) bgpValidationRules(config map[string]string) (map[string]func(value string) error, error) {
 	rules := map[string]func(value string) error{}
 	for k := range config {
-		// BGP keys have the peer name in their name, extract the suffix.
+		// only BGP keys are handled here.
 		if !strings.HasPrefix(k, "bgp.") {
```
Comment (roosterfish) [inline lxd/network/driver_common.go:582]: Thanks @tomponline for the clarification. 

@mandrav then I would suggest to revert the changes (you can leave the `%q` for the error message) and extend this line to just check for `bgp.peers.` which will exclude the two nexthop settings.

```suggestion
		if !strings.HasPrefix(k, "bgp.peers.") {
```

**Review by roosterfish — CHANGES_REQUESTED** [review]

**Review by roosterfish — APPROVED** [review]

**Review by tomponline — APPROVED** [review]
